python/pyarrow/tests/test_io.py

```diff
@@ -586,6 +741,7 @@ def sample_disk_data(request, tmpdir):
 
     def teardown():
         _try_delete(path)
+
```

Comment:
It strikes me that you added a lot of tests for `PythonFile` but none for `NativeFile` (which is the focus of this PR). Did you forget them?

Reply:
`NativeFile` ist kind of an abstract base and cannot used directly (it doesn't set rd_file/wr_file, for example). `PythonFile` is a close descendant which just specifies a custom initializer. Looking at the existing tests, non seem to test `NativeFile` directly.

Reply:
Yes, the purpose of NativeFile is to provide the abstract interfaces to say "I have a C++ object implementing one of the Arrow C++ IO interfaces" so that C++ code can acquire the internals of a Python-created file resource and use it without interacting with the GIL, or having to pass through PyBytes object in some cases